File: configuration-manager-go-client/settings_service.go
package configmanager

import (
	"context"
	"fmt"
)

// MongoDB Settings Methods

// GetMongoDBSettings gets MongoDB connection settings and status (admin only)
func (s *SettingsService) GetMongoDBSettings(ctx context.Context) (*MongoDBSettingsResponse, error) {
	resp, err := s.client.Get(ctx, "/settings/mongodb")
	if err != nil {
		return nil, fmt.Errorf("get MongoDB settings request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MongoDBSettingsResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode MongoDB settings response: %w", err)
	}

	return &response, nil
}

// UpdateMongoDBSettings updates MongoDB connection settings (admin only)
func (s *SettingsService) UpdateMongoDBSettings(ctx context.Context, settings *MongoDBSettings) (*MigrationResponse, error) {
	if settings == nil {
		return nil, NewValidationError("MongoDB settings cannot be nil", 0)
	}

	if settings.ConnectionString == "" {
		return nil, NewValidationError("connection string cannot be empty", 0)
	}

	resp, err := s.client.Put(ctx, "/settings/mongodb", settings)
	if err != nil {
		return nil, fmt.Errorf("update MongoDB settings request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MigrationResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode update MongoDB settings response: %w", err)
	}

	return &response, nil
}

// TestMongoDBConnection tests MongoDB connection (admin only)
func (s *SettingsService) TestMongoDBConnection(ctx context.Context, request *MongoDBTestRequest) (*MongoDBTestResponse, error) {
	if request == nil {
		return nil, NewValidationError("MongoDB test request cannot be nil", 0)
	}

	if request.ConnectionString == "" {
		return nil, NewValidationError("connection string cannot be empty", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/mongodb/test", request)
	if err != nil {
		return nil, fmt.Errorf("test MongoDB connection request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MongoDBTestResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode test MongoDB connection response: %w", err)
	}

	return &response, nil
}

// ConnectToMongoDB connects to MongoDB (admin only)
func (s *SettingsService) ConnectToMongoDB(ctx context.Context, settings *MongoDBSettings) (*MongoDBTestResponse, error) {
	if settings == nil {
		return nil, NewValidationError("MongoDB settings cannot be nil", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/mongodb/connect", settings)
	if err != nil {
		return nil, fmt.Errorf("connect to MongoDB request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MongoDBTestResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode connect to MongoDB response: %w", err)
	}

	return &response, nil
}

// DisconnectFromMongoDB disconnects from MongoDB (admin only)
func (s *SettingsService) DisconnectFromMongoDB(ctx context.Context) (*MongoDBTestResponse, error) {
	resp, err := s.client.Post(ctx, "/settings/mongodb/disconnect", nil)
	if err != nil {
		return nil, fmt.Errorf("disconnect from MongoDB request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MongoDBTestResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode disconnect from MongoDB response: %w", err)
	}

	return &response, nil
}

// MigrateToMongoDB migrates data from SQLite to MongoDB (admin only)
func (s *SettingsService) MigrateToMongoDB(ctx context.Context, request *MigrationRequest) (*MigrationResponse, error) {
	if request == nil {
		return nil, NewValidationError("migration request cannot be nil", 0)
	}

	if request.ConnectionString == "" {
		return nil, NewValidationError("connection string cannot be empty", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/mongodb/migrate", request)
	if err != nil {
		return nil, fmt.Errorf("migrate to MongoDB request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MigrationResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode migrate to MongoDB response: %w", err)
	}

	return &response, nil
}

// GetMongoDBStatus gets MongoDB connection status (admin only)
func (s *SettingsService) GetMongoDBStatus(ctx context.Context) (*MongoDBStatus, error) {
	resp, err := s.client.Get(ctx, "/settings/mongodb/status")
	if err != nil {
		return nil, fmt.Errorf("get MongoDB status request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MongoDBStatus
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode MongoDB status response: %w", err)
	}

	return &response, nil
}

// MigrateToEmbeddedMongoDB migrates to embedded MongoDB (admin only)
func (s *SettingsService) MigrateToEmbeddedMongoDB(ctx context.Context) (*MigrationResponse, error) {
	resp, err := s.client.Post(ctx, "/settings/mongodb/migrate-embedded", nil)
	if err != nil {
		return nil, fmt.Errorf("migrate to embedded MongoDB request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MigrationResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode migrate to embedded MongoDB response: %w", err)
	}

	return &response, nil
}

// RevertToSQLite reverts to SQLite from MongoDB (admin only)
func (s *SettingsService) RevertToSQLite(ctx context.Context, request *RevertToSQLiteRequest) (*MigrationResponse, error) {
	if request == nil {
		return nil, NewValidationError("revert to SQLite request cannot be nil", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/mongodb/revert-to-sqlite", request)
	if err != nil {
		return nil, fmt.Errorf("revert to SQLite request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MigrationResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode revert to SQLite response: %w", err)
	}

	return &response, nil
}

// Data Management Methods

// GetDataStatistics gets data statistics (admin only)
func (s *SettingsService) GetDataStatistics(ctx context.Context) (*DataStatisticsResponse, error) {
	resp, err := s.client.Get(ctx, "/settings/data/status")
	if err != nil {
		return nil, fmt.Errorf("get data statistics request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response DataStatisticsResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode data statistics response: %w", err)
	}

	return &response, nil
}

// CreateBackup creates data backup (admin only)
func (s *SettingsService) CreateBackup(ctx context.Context, request *CreateBackupRequest) (*MigrationResponse, error) {
	if request == nil {
		return nil, NewValidationError("create backup request cannot be nil", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/data/backup", request)
	if err != nil {
		return nil, fmt.Errorf("create backup request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response MigrationResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode create backup response: %w", err)
	}

	return &response, nil
}

// GetBackups lists available backups (admin only)
func (s *SettingsService) GetBackups(ctx context.Context) (*BackupListResponse, error) {
	resp, err := s.client.Get(ctx, "/settings/data/backups")
	if err != nil {
		return nil, fmt.Errorf("get backups request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response BackupListResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode backups response: %w", err)
	}

	return &response, nil
}

// RestoreFromBackup restores from backup (admin only)
func (s *SettingsService) RestoreFromBackup(ctx context.Context, request *RestoreRequest) (*RestoreResponse, error) {
	if request == nil {
		return nil, NewValidationError("restore request cannot be nil", 0)
	}

	if request.BackupName == "" {
		return nil, NewValidationError("backup name cannot be empty", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/data/restore", request)
	if err != nil {
		return nil, fmt.Errorf("restore from backup request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response RestoreResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode restore from backup response: %w", err)
	}

	return &response, nil
}

// Storage Settings Methods

// GetStorageSettings gets storage settings (admin only)
func (s *SettingsService) GetStorageSettings(ctx context.Context) (*StorageSettingsResponse, error) {
	resp, err := s.client.Get(ctx, "/settings/storage")
	if err != nil {
		return nil, fmt.Errorf("get storage settings request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response StorageSettingsResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode storage settings response: %w", err)
	}

	return &response, nil
}

// UpdateStorageSettings updates storage settings (admin only)
func (s *SettingsService) UpdateStorageSettings(ctx context.Context, settings *StorageSettings) (*StorageTestResponse, error) {
	if settings == nil {
		return nil, NewValidationError("storage settings cannot be nil", 0)
	}

	// Validate S3 settings if storage type is S3
	if settings.StorageType == StorageTypeS3 {
		if settings.S3BucketName == nil || *settings.S3BucketName == "" {
			return nil, NewValidationError("S3 bucket name is required for S3 storage", 0)
		}

		if settings.AWSRegion == nil || *settings.AWSRegion == "" {
			return nil, NewValidationError("AWS region is required for S3 storage", 0)
		}

		if settings.AWSAccessKeyID == nil || *settings.AWSAccessKeyID == "" {
			return nil, NewValidationError("AWS access key ID is required for S3 storage", 0)
		}

		if settings.AWSSecretAccessKey == nil || *settings.AWSSecretAccessKey == "" {
			return nil, NewValidationError("AWS secret access key is required for S3 storage", 0)
		}
	}

	resp, err := s.client.Put(ctx, "/settings/storage", settings)
	if err != nil {
		return nil, fmt.Errorf("update storage settings request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response StorageTestResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode update storage settings response: %w", err)
	}

	return &response, nil
}

// TestStorageConnection tests storage connection (admin only)
func (s *SettingsService) TestStorageConnection(ctx context.Context, settings *StorageSettings) (*StorageTestResponse, error) {
	if settings == nil {
		return nil, NewValidationError("storage settings cannot be nil", 0)
	}

	resp, err := s.client.Post(ctx, "/settings/storage/test", settings)
	if err != nil {
		return nil, fmt.Errorf("test storage connection request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response StorageTestResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode test storage connection response: %w", err)
	}

	return &response, nil
}

// GetStorageStatus gets storage status (admin only)
func (s *SettingsService) GetStorageStatus(ctx context.Context) (*StorageStatusResponse, error) {
	resp, err := s.client.Get(ctx, "/settings/storage/status")
	if err != nil {
		return nil, fmt.Errorf("get storage status request failed: %w", err)
	}
	defer s.client.CloseResponse(resp)

	var response StorageStatusResponse
	if err := s.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode storage status response: %w", err)
	}

	return &response, nil
}

File: configuration-manager-python-client/configuration_manager_client/services/settings_service.py
"""Settings and administration service for Configuration Manager API."""

from typing import Optional

from ..http_client import HTTPClient
from ..models.settings import (
    MongoDBSettings,
    MongoDBSettingsResponse,
    MongoDBTestRequest,
    MongoDBTestResponse,
    MongoDBStatus,
    MigrationRequest,
    MigrationResponse,
    DataStatisticsResponse,
    CreateBackupRequest,
    BackupListResponse,
    RestoreRequest,
    RestoreResponse,
    StorageSettings,
    StorageSettingsResponse,
    StorageTestResponse,
    StorageStatusResponse,
    RevertToSQLiteRequest,
)
from ..models.types import StorageType
from ..exceptions import ValidationError
from ..utils import from_dict


class SettingsService:
    """Settings and administration service client."""

    def __init__(self, http_client: HTTPClient):
        """Initialize settings service.
        
        Args:
            http_client: HTTP client instance
        """
        self.client = http_client

    # MongoDB Settings Methods

    def get_mongodb_settings(self) -> MongoDBSettingsResponse:
        """Get MongoDB connection settings and status (admin only).
        
        Returns:
            MongoDBSettingsResponse containing settings and status
        """
        response = self.client.get("/settings/mongodb")
        return from_dict(MongoDBSettingsResponse, response.json())

    def update_mongodb_settings(self, settings: MongoDBSettings) -> MigrationResponse:
        """Update MongoDB connection settings (admin only).
        
        Args:
            settings: MongoDB settings
            
        Returns:
            MigrationResponse containing update result
            
        Raises:
            ValidationError: If settings are invalid
        """
        if not settings.connection_string:
            raise ValidationError("Connection string cannot be empty")

        data = {
            "connectionString": settings.connection_string,
        }

        if settings.options:
            data["options"] = {
                "useNewUrlParser": settings.options.use_new_url_parser,
                "useUnifiedTopology": settings.options.use_unified_topology,
            }

        response = self.client.put("/settings/mongodb", json_data=data)
        return from_dict(MigrationResponse, response.json())

    def test_mongodb_connection(self, request: MongoDBTestRequest) -> MongoDBTestResponse:
        """Test MongoDB connection (admin only).
        
        Args:
            request: MongoDB test request
            
        Returns:
            MongoDBTestResponse containing test result
            
        Raises:
            ValidationError: If request is invalid
        """
        if not request.connection_string:
            raise ValidationError("Connection string cannot be empty")

        data = {"connectionString": request.connection_string}
        response = self.client.post("/settings/mongodb/test", json_data=data)
        return from_dict(MongoDBTestResponse, response.json())

    def connect_to_mongodb(self, settings: MongoDBSettings) -> MongoDBTestResponse:
        """Connect to MongoDB (admin only).
        
        Args:
            settings: MongoDB settings
            
        Returns:
            MongoDBTestResponse containing connection result
        """
        data = {
            "connectionString": settings.connection_string,
        }

        if settings.options:
            data["options"] = {
                "useNewUrlParser": settings.options.use_new_url_parser,
                "useUnifiedTopology": settings.options.use_unified_topology,
            }

        response = self.client.post("/settings/mongodb/connect", json_data=data)
        return from_dict(MongoDBTestResponse, response.json())

    def disconnect_from_mongodb(self) -> MongoDBTestResponse:
        """Disconnect from MongoDB (admin only).
        
        Returns:
            MongoDBTestResponse containing disconnection result
        """
        response = self.client.post("/settings/mongodb/disconnect")
        return from_dict(MongoDBTestResponse, response.json())

    def migrate_to_mongodb(self, request: MigrationRequest) -> MigrationResponse:
        """Migrate data from SQLite to MongoDB (admin only).
        
        Args:
            request: Migration request
            
        Returns:
            MigrationResponse containing migration result
            
        Raises:
            ValidationError: If request is invalid
        """
        if not request.connection_string:
            raise ValidationError("Connection string cannot be empty")

        data = {
            "connectionString": request.connection_string,
        }

        if request.create_backup is not None:
            data["createBackup"] = request.create_backup

        response = self.client.post("/settings/mongodb/migrate", json_data=data)
        return from_dict(MigrationResponse, response.json())

    def get_mongodb_status(self) -> MongoDBStatus:
        """Get MongoDB connection status (admin only).
        
        Returns:
            MongoDBStatus containing connection status
        """
        response = self.client.get("/settings/mongodb/status")
        return from_dict(MongoDBStatus, response.json())

    def migrate_to_embedded_mongodb(self) -> MigrationResponse:
        """Migrate to embedded MongoDB (admin only).
        
        Returns:
            MigrationResponse containing migration result
        """
        response = self.client.post("/settings/mongodb/migrate-embedded")
        return from_dict(MigrationResponse, response.json())

    def revert_to_sqlite(self, request: Optional[RevertToSQLiteRequest] = None) -> MigrationResponse:
        """Revert to SQLite from MongoDB (admin only).
        
        Args:
            request: Optional revert request
            
        Returns:
            MigrationResponse containing revert result
        """
        data = {}
        if request and request.migrate_data is not None:
            data["migrateData"] = request.migrate_data

        response = self.client.post("/settings/mongodb/revert-to-sqlite", json_data=data)
        return from_dict(MigrationResponse, response.json())

    # Data Management Methods

    def get_data_statistics(self) -> DataStatisticsResponse:
        """Get data statistics (admin only).
        
        Returns:
            DataStatisticsResponse containing system statistics
        """
        response = self.client.get("/settings/data/status")
        return from_dict(DataStatisticsResponse, response.json())

    def create_backup(self, request: Optional[CreateBackupRequest] = None) -> MigrationResponse:
        """Create data backup (admin only).
        
        Args:
            request: Optional backup request with name
            
        Returns:
            MigrationResponse containing backup result
        """
        data = {}
        if request and request.name:
            data["name"] = request.name

        response = self.client.post("/settings/data/backup", json_data=data)
        return from_dict(MigrationResponse, response.json())

    def get_backups(self) -> BackupListResponse:
        """List available backups (admin only).
        
        Returns:
            BackupListResponse containing list of backups
        """
        response = self.client.get("/settings/data/backups")
        return from_dict(BackupListResponse, response.json())

    def restore_from_backup(self, request: RestoreRequest) -> RestoreResponse:
        """Restore from backup (admin only).
        
        Args:
            request: Restore request
            
        Returns:
            RestoreResponse containing restore result
            
        Raises:
            ValidationError: If backup name is empty
        """
        if not request.backup_name:
            raise ValidationError("Backup name cannot be empty")

        data = {
            "backupName": request.backup_name,
        }

        if request.create_backup is not None:
            data["createBackup"] = request.create_backup

        response = self.client.post("/settings/data/restore", json_data=data)
        return from_dict(RestoreResponse, response.json())

    # Storage Settings Methods

    def get_storage_settings(self) -> StorageSettingsResponse:
        """Get storage settings (admin only).
        
        Returns:
            StorageSettingsResponse containing storage configuration
        """
        response = self.client.get("/settings/storage")
        return from_dict(StorageSettingsResponse, response.json())

    def update_storage_settings(self, settings: StorageSettings) -> StorageTestResponse:
        """Update storage settings (admin only).
        
        Args:
            settings: Storage settings
            
        Returns:
            StorageTestResponse containing update result
            
        Raises:
            ValidationError: If settings are invalid
        """
        # Validate S3 settings if storage type is S3
        if settings.storage_type == StorageType.S3:
            if not settings.s3_bucket_name:
                raise ValidationError("S3 bucket name is required for S3 storage")

            if not settings.aws_region:
                raise ValidationError("AWS region is required for S3 storage")

            if not settings.aws_access_key_id:
                raise ValidationError("AWS access key ID is required for S3 storage")

            if not settings.aws_secret_access_key:
                raise ValidationError("AWS secret access key is required for S3 storage")

        data = {
            "storageType": settings.storage_type.value,
        }

        if settings.s3_bucket_name:
            data["s3BucketName"] = settings.s3_bucket_name

        if settings.aws_region:
            data["awsRegion"] = settings.aws_region

        if settings.aws_access_key_id:
            data["awsAccessKeyId"] = settings.aws_access_key_id

        if settings.aws_secret_access_key:
            data["awsSecretAccessKey"] = settings.aws_secret_access_key

        response = self.client.put("/settings/storage", json_data=data)
        return from_dict(StorageTestResponse, response.json())

    def test_storage_connection(self, settings: StorageSettings) -> StorageTestResponse:
        """Test storage connection (admin only).
        
        Args:
            settings: Storage settings to test
            
        Returns:
            StorageTestResponse containing test result
        """
        data = {
            "storageType": settings.storage_type.value,
        }

        if settings.s3_bucket_name:
            data["s3BucketName"] = settings.s3_bucket_name

        if settings.aws_region:
            data["awsRegion"] = settings.aws_region

        if settings.aws_access_key_id:
            data["awsAccessKeyId"] = settings.aws_access_key_id

        if settings.aws_secret_access_key:
            data["awsSecretAccessKey"] = settings.aws_secret_access_key

        response = self.client.post("/settings/storage/test", json_data=data)
        return from_dict(StorageTestResponse, response.json())

    def get_storage_status(self) -> StorageStatusResponse:
        """Get storage status (admin only).
        
        Returns:
            StorageStatusResponse containing storage status
        """
        response = self.client.get("/settings/storage/status")
        return from_dict(StorageStatusResponse, response.json())
